server/src/main/java/io/druid/segment/indexing/TransformSpec.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.indexing;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+import io.druid.data.input.impl.InputRowParser;
+import io.druid.data.input.impl.StringInputRowParser;
+import io.druid.java.util.common.ISE;
+import io.druid.query.filter.DimFilter;
+
+import javax.annotation.Nullable;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+public class TransformSpec
+{
+  public static final TransformSpec NONE = new TransformSpec(null, null);
+
+  private final DimFilter filter;
+  private final List<Transform> transforms;
+
+  @JsonCreator
+  public TransformSpec(
+      @JsonProperty("filter") final DimFilter filter,
+      @JsonProperty("transforms") final List<Transform> transforms
+  )
+  {
+    this.filter = filter;
+    this.transforms = transforms == null ? ImmutableList.of() : transforms;
+
+    // Check for name collisions.
+    final Set<String> seen = new HashSet<>();
+    for (Transform transform : this.transforms) {
+      if (!seen.add(transform.getName())) {
+        throw new ISE("Transform name '%s' cannot be used twice", transform.getName());
+      }
+    }
+  }
+
+  public static <T> TransformSpec fromInputRowParser(final InputRowParser<T> parser)
+  {
+    // Hack: some firehoses and input specs must extract transformSpec from the parser, since they do not
+    // actually use the parser, but still must respect the transformSpec. This method should extract whatever
+    // transformSpec "decorate" had put in.
+
+    if (parser instanceof TransformingInputRowParser) {
+      return ((TransformingInputRowParser) parser).getTransformSpec();
+    } else if (parser instanceof TransformingStringInputRowParser) {
+      return ((TransformingStringInputRowParser) parser).getTransformSpec();
+    } else {
+      return TransformSpec.NONE;
```

Comment:
should we throw a RuntimeException in this case because this should only get called in the codepaths where a `decorate(..)` would've happened. If we reached here that means there is some coding error ?

Reply:
This case can happen if the transformSpec is a no-op, in which case `decorate` will return the original parser, not a decorated one.

Reply:
ok, this makes coding errors harder to catch though. if we always did the wrapping (costs only one object) in `decorate` and threw exception here then catching coding errors would be trivial.

but, at this point its more of a nit than anything so feel free to ignore.

Reply:
Hmm. I had done it this way so that if you're not using transformSpec, it has no effect (parser isn't wrapped), mostly so just in case there was some bug with decorated parsers it won't affect all users. Although I could have it always wrap the parser so bugs (if any) are found more quickly.

What do you think?